In [1]:
import pandas as pd                 #(for dataframes)
import numpy as np                 #(for numerical computation)
import matplotlib.pyplot as plt    #(for ploting)
import seaborn as sns               #(for data visuvalization)
import plotly.express as px         #(for interactive scatter plot)
import plotly.io as pio         #for configure how plots are displayed 
                                 #(e.g., in notebooks, browsers, or exported))

pio.renderers.default = "svg"    #(it will render as an SVG image by default)

In [2]:
#!pip install plotly


In [3]:
raw_df=pd.read_csv('Zomato_Mumbai_Dataset.csv',delimiter='|') 
#(delimiter - split each row into columns wherever you see a pipe symbol)

In [4]:
raw_df.head() #(first 5 rows)

NAME PRICE                                    CUSINE_CATEGORY  \
0         Hitchki  1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda   400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu  1800                                      Asian,Chinese   
3  Butterfly High  1000                                      Modern Indian   
4       BKC  DIVE  1200                   North Indian,Chinese,Continental   

     CITY                                             REGION  \
0  Mumbai  First International Financial Centre-- Bandra ...   
1  Mumbai                                              Mahim   
2  Mumbai                                               Juhu   
3  Mumbai                               Bandra Kurla Complex   
4  Mumbai                               Bandra Kurla Complex   

                                                 URL PAGE NO     CUSINE TYPE  \
0  https://www.zomato.com/mumbai/hitchki-bandra-k...       1   Casual Dining   
1  https://www.zomato.com/mumbai/baba-falooda-mah...       1  Dessert Parlor   
2  https://www.zomato.com/mumbai/chin-chin-chu-ju...       1   Casual Dining   
3  https://www.zomato.com/mumbai/butterfly-high-b...       1             Bar   
4  https://www.zomato.com/mumbai/bkc-dive-bandra-...       1             Bar   

                     TIMING  RATING_TYPE RATING VOTES  
0  12noon to 130am(Mon-Sun)    Excellent    4.9  3529  
1       2pm to 1am(Mon-Sun)    Very Good    4.4  1723  
2    12noon to 1am(Mon-Sun)    Very Good    4.2   337  
3  12noon to 130am(Mon-Sun)    Very Good    4.3  1200  
4    1130am to 1am(Mon-Sun)  Veľmi dobré    4.4  5995

# 3.Getting basic information about the Dataset

In [5]:
raw_df.shape   #(here we can know about total column and row in this case 15081 rows and 12 columns)

(15081, 12)

In [6]:
raw_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15081 entries, 0 to 15080
Data columns (total 12 columns):
NAME               15081 non-null object
PRICE              15080 non-null object
CUSINE_CATEGORY    15079 non-null object
CITY               15080 non-null object
REGION             15080 non-null object
URL                15080 non-null object
PAGE NO            15080 non-null object
CUSINE TYPE        15080 non-null object
TIMING             15015 non-null object
RATING_TYPE        15080 non-null object
RATING             15080 non-null object
VOTES              15080 non-null object
dtypes: object(12)
memory usage: 1.4+ MB


In [7]:
raw_df.describe()  #(in this table we can find how many dublicates entries are there among total count)

NAME  PRICE  CUSINE_CATEGORY    CITY  REGION    URL  PAGE NO  \
count   15081  15080            15079   15080   15080  15080    15080   
unique  12720     67             3183       2     241  13823      944   
top      NAME    400  CUSINE_CATEGORY  Mumbai  REGION    URL  PAGE NO   
freq      942   2042              942   14138     942    942      942   

        CUSINE TYPE                 TIMING RATING_TYPE RATING  VOTES  
count         15080                  15015       15080  15080  15080  
unique           23                   2551          32     35   1124  
top     Quick Bites  11am to 11pm(Mon-Sun)     Average      -      -  
freq           5262                   1192        5112   2360   2360

# 4. cleaning the dataset

#removing dublicate rows of dataset


#check for dublicates 

In [8]:
wrong_data=raw_df['PAGE NO']=='PAGE NO'

In [9]:
wrong_data

0        False
1        False
2        False
3        False
4        False
         ...  
15076    False
15077    False
15078    False
15079    False
15080    False
Name: PAGE NO, Length: 15081, dtype: bool

In [10]:
raw_df[wrong_data]

NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE  \
15     NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
31     NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
47     NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
63     NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
79     NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
...     ...    ...              ...   ...     ...  ...      ...          ...   
15000  NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
15016  NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
15032  NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
15048  NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   
15064  NAME  PRICE  CUSINE_CATEGORY  CITY  REGION  URL  PAGE NO  CUSINE TYPE   

       TIMING  RATING_TYPE  RATING  VOTES  
15     TIMING  RATING_TYPE  RATING  VOTES  
31     TIMING  RATING_TYPE  RATING  VOTES  
47     TIMING  RATING_TYPE  RATING  VOTES  
63     TIMING  RATING_TYPE  RATING  VOTES  
79     TIMING  RATING_TYPE  RATING  VOTES  
...       ...          ...     ...    ...  
15000  TIMING  RATING_TYPE  RATING  VOTES  
15016  TIMING  RATING_TYPE  RATING  VOTES  
15032  TIMING  RATING_TYPE  RATING  VOTES  
15048  TIMING  RATING_TYPE  RATING  VOTES  
15064  TIMING  RATING_TYPE  RATING  VOTES  

[942 rows x 12 columns]

In [11]:
#fliping the true values of the wrong data and then sorting the corrrect databack in the raw_df dataframe 
#this permantly remove the wrong data from the origional dataframe
# why fliping wrong dataset is nessasary?
## here as a bool wrong (True) and valid (False) 
## So we have to flip right? 
## thats why we use ~ (tilde operator) this is for Remove all duplicate only keep unique values

In [12]:
raw_df=raw_df[~wrong_data]

In [13]:
raw_df

NAME PRICE  \
0                Hitchki  1200   
1           Baba Falooda   400   
2          Chin Chin Chu  1800   
3         Butterfly High  1000   
4              BKC  DIVE  1200   
...                  ...   ...   
15076  Hari Om Snack Bar   350   
15077           PitaBurg   400   
15078     Uncha Otlawala   300   
15079     Mandarin Panda   400   
15080                     NaN   

                                         CUSINE_CATEGORY    CITY  \
0      Modern Indian,North Indian,Chinese,Momos,Birya...  Mumbai   
1                           Desserts,Ice Cream,Beverages  Mumbai   
2                                          Asian,Chinese  Mumbai   
3                                          Modern Indian  Mumbai   
4                       North Indian,Chinese,Continental  Mumbai   
...                                                  ...     ...   
15076                     Fast Food,South Indian,Chinese  Mumbai   
15077                                 Fast Food,Lebanese  Mumbai   
15078                                 Desserts,Ice Cream  Mumbai   
15079                              Desserts,Chinese,Thai  Mumbai   
15080                                                NaN     NaN   

                                                  REGION  \
0      First International Financial Centre-- Bandra ...   
1                                                  Mahim   
2                                                   Juhu   
3                                   Bandra Kurla Complex   
4                                   Bandra Kurla Complex   
...                                                  ...   
15076                                     Kandivali West   
15077                                        Lower Parel   
15078                                     Kandivali West   
15079                                         Malad West   
15080                                                NaN   

                                                     URL PAGE NO  \
0      https://www.zomato.com/mumbai/hitchki-bandra-k...       1   
1      https://www.zomato.com/mumbai/baba-falooda-mah...       1   
2      https://www.zomato.com/mumbai/chin-chin-chu-ju...       1   
3      https://www.zomato.com/mumbai/butterfly-high-b...       1   
4      https://www.zomato.com/mumbai/bkc-dive-bandra-...       1   
...                                                  ...     ...   
15076  https://www.zomato.com/mumbai/hari-om-snack-ba...      99   
15077  https://www.zomato.com/mumbai/pitaburg-lower-p...      99   
15078  https://www.zomato.com/mumbai/uncha-otlawala-1...      99   
15079  https://www.zomato.com/mumbai/mandarin-panda-m...      99   
15080                                                NaN     NaN   

          CUSINE TYPE                                         TIMING  \
0       Casual Dining                       12noon to 130am(Mon-Sun)   
1      Dessert Parlor                            2pm to 1am(Mon-Sun)   
2       Casual Dining                         12noon to 1am(Mon-Sun)   
3                 Bar                       12noon to 130am(Mon-Sun)   
4                 Bar                         1130am to 1am(Mon-Sun)   
...               ...                                            ...   
15076     Quick Bites                         11am to 230am(Mon-Sun)   
15077            none  11am to 11pm(Mon,Tue,Wed,Thu,Sun),11am to ...   
15078  Dessert Parlor                         9am to 1230AM(Mon-Sun)   
15079            none            12noon to 330pm,7pm to 1am(Mon-Sun)   
15080             NaN                                            NaN   

       RATING_TYPE RATING VOTES  
0        Excellent    4.9  3529  
1        Very Good    4.4  1723  
2        Very Good    4.2   337  
3        Very Good    4.3  1200  
4      Veľmi dobré    4.4  5995  
...            ...    ...   ...  
15076         Good    3.7    64  
15077      Average    3.4    99  
15078         Good    3.5    29  
15079         Good    3.7   121  
15080          Na

In [14]:
## Dropping columns which are not required for further analsis
raw_df.drop(['URL','PAGE NO','CITY'],axis=1,inplace=True)

In [15]:
raw_df.head()

NAME PRICE                                    CUSINE_CATEGORY  \
0         Hitchki  1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda   400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu  1800                                      Asian,Chinese   
3  Butterfly High  1000                                      Modern Indian   
4       BKC  DIVE  1200                   North Indian,Chinese,Continental   

                                              REGION     CUSINE TYPE  \
0  First International Financial Centre-- Bandra ...   Casual Dining   
1                                              Mahim  Dessert Parlor   
2                                               Juhu   Casual Dining   
3                               Bandra Kurla Complex             Bar   
4                               Bandra Kurla Complex             Bar   

                     TIMING  RATING_TYPE RATING VOTES  
0  12noon to 130am(Mon-Sun)    Excellent    4.9  3529  
1       2pm to 1am(Mon-Sun)    Very Good    4.4  1723  
2    12noon to 1am(Mon-Sun)    Very Good    4.2   337  
3  12noon to 130am(Mon-Sun)    Very Good    4.3  1200  
4    1130am to 1am(Mon-Sun)  Veľmi dobré    4.4  5995

In [16]:
## Removing the null records
raw_df.isnull().sum()

NAME                0
PRICE               1
CUSINE_CATEGORY     2
REGION              1
CUSINE TYPE         1
TIMING             66
RATING_TYPE         1
RATING              1
VOTES               1
dtype: int64

In [17]:
## checking for null rows
raw_df[raw_df['PRICE'].isnull()]

NAME PRICE CUSINE_CATEGORY REGION CUSINE TYPE TIMING RATING_TYPE RATING  \
15080       NaN             NaN    NaN         NaN    NaN         NaN    NaN   

      VOTES  
15080   NaN

In [18]:
## Drop the above NaN row from the raw_df
raw_df=raw_df.drop(labels=15080,axis=0)


In [19]:
## Replace the null(missing values) values with NA (not available)
raw_df.fillna('NA',inplace=True)


In [20]:
## check for null values
raw_df.isnull().sum()

NAME               0
PRICE              0
CUSINE_CATEGORY    0
REGION             0
CUSINE TYPE        0
TIMING             0
RATING_TYPE        0
RATING             0
VOTES              0
dtype: int64

In [21]:
## Now our dataset has no null values but we have to check for datatype 
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14138 entries, 0 to 15079
Data columns (total 9 columns):
NAME               14138 non-null object
PRICE              14138 non-null object
CUSINE_CATEGORY    14138 non-null object
REGION             14138 non-null object
CUSINE TYPE        14138 non-null object
TIMING             14138 non-null object
RATING_TYPE        14138 non-null object
RATING             14138 non-null object
VOTES              14138 non-null object
dtypes: object(9)
memory usage: 1.1+ MB


In [22]:
## Here all datatypes are object but we have numerical columns too 
## so we should convert the datatype of numerical columns to numeric datatype.
## checking for text values in the column before converting it to numeric datatype.
raw_df['RATING'].value_counts()

-          2360
3.5        1094
3.4        1036
3.6         960
NEW         953
3.3         926
3.7         917
3.2         801
3.8         782
3.1         734
3.0         622
3.9         596
2.9         409
4.0         408
2.8         309
4.1         298
4.2         199
2.7         170
4.3         148
4.4          99
2.6          77
Opening      57
4.5          46
2.5          39
4.6          32
2.4          26
4.7          13
2.3          10
2.1           5
2.2           4
4.8           4
4.9           2
2.0           1
1.8           1
Name: RATING, dtype: int64

In [23]:
## Replace the string (-,New,Opening) with "0"
## why rating should be in the numerical column so we have to change all string into numerical
raw_df['RATING'].replace(to_replace=['-','NEW','Opening'],value='0',inplace=True)

In [24]:
## checking for Price column 
raw_df['PRICE'].value_counts()

400     2042
500     1954
300     1936
200     1177
600     1062
        ... 
80         1
2600       1
5          1
1450       1
580        1
Name: PRICE, Length: 66, dtype: int64

In [25]:
raw_df['PRICE'].dtype


dtype('O')

## Replace strings in Price column to convert it into numerical

import pandas as pd
pd.set_option('display.max_seq_items', None)
raw_df['PRICE'].unique()

## OUtput values numbers are in string so that it shows object.
array(['1200', '400', '1800', '1000', '800', '1300', '1500', '600',
       '1400', '1100', '2000', '350', '900', '700', '500', '1600', '150',
       '300', '550', '450', '650', '750', '250', '200', '850', '100',
       '2500', '1450', '180', '950', '1700', '1900', '2600', '2400',
       '3000', '1250', '3200', '505', '2200', '1050', '1650', '1350',
       '50', '920', '1150', '1550', '3500', '398', '5000', '120', '220',
       '160', '340', '5', '480', '2100', '580', '1850', '249', '80',
       '2300', '0', '360', '3502', '248', '2700'], dtype=object)
       
       In this case we just convet the dtype from object to int thats enough.

In [26]:
## checking for votes column 
raw_df['VOTES'].value_counts()

-       2360
NEW      953
4        364
5        320
6        288
        ... 
1225       1
1541       1
381        1
2224       1
492        1
Name: VOTES, Length: 1123, dtype: int64

In [27]:
## Replace strings in VOTES column to convert it into numerical
raw_df['VOTES'].replace(to_replace=['-','NEW','Opening'],value='0',inplace=True)

In [28]:
## Changing datatype of the numerical columns as int or float 
raw_df['PRICE']= raw_df['PRICE'].astype('int64')
raw_df['VOTES']= raw_df['VOTES'].astype('int64')
raw_df['RATING'] = raw_df['RATING'].astype('float64') 

In [29]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14138 entries, 0 to 15079
Data columns (total 9 columns):
NAME               14138 non-null object
PRICE              14138 non-null int64
CUSINE_CATEGORY    14138 non-null object
REGION             14138 non-null object
CUSINE TYPE        14138 non-null object
TIMING             14138 non-null object
RATING_TYPE        14138 non-null object
RATING             14138 non-null float64
VOTES              14138 non-null int64
dtypes: float64(1), int64(2), object(6)
memory usage: 1.1+ MB


In [30]:
## working with timing 
raw_df['TIMING'].value_counts()

11am to 11pm(Mon-Sun)                              1192
11am to 12midnight(Mon-Sun)                         632
12noon to 12midnight(Mon-Sun)                       467
11am to 1130pm(Mon-Sun)                             309
10am to 10pm(Mon-Sun)                               267
                                                   ... 
Closed(Mon),10am to 11pm(Tue-Thu),10am to ...         1
1130am to 4pm(Mon-Sun)                                1
1130am to 1130pm(Mon,Tue,Wed,Thu,Fri,Sun)...          1
1230pm to 4pm,7pm to 11pm(Mon-Sun)                    1
8am to 11pm(Mon,Tue,Fri,Sat,Sun),6pm to 11pm...       1
Name: TIMING, Length: 2551, dtype: int64

In [31]:
## split the columns and storing it in tem_df temporary dataframe
tem_df=raw_df['TIMING'].str.split("(",n=1,expand=True)

In [32]:
tem_df

0                                 1
0                      12noon to 130am                          Mon-Sun)
1                           2pm to 1am                          Mon-Sun)
2                        12noon to 1am                          Mon-Sun)
3                      12noon to 130am                          Mon-Sun)
4                        1130am to 1am                          Mon-Sun)
...                                ...                               ...
15075  8am to 11pm,12midnight to 115am                          Mon-Sun)
15076                    11am to 230am                          Mon-Sun)
15077                     11am to 11pm  Mon,Tue,Wed,Thu,Sun),11am to ...
15078                    9am to 1230AM                          Mon-Sun)
15079       12noon to 330pm,7pm to 1am                          Mon-Sun)

[14138 rows x 2 columns]

In [33]:
## Assigning the columns back to the raw_df dataframe
raw_df['TIMING']=tem_df[0]
raw_df['DAYS_OPENS']=tem_df[1]
raw_df.head()

NAME  PRICE                                    CUSINE_CATEGORY  \
0         Hitchki   1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda    400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu   1800                                      Asian,Chinese   
3  Butterfly High   1000                                      Modern Indian   
4       BKC  DIVE   1200                   North Indian,Chinese,Continental   

                                              REGION     CUSINE TYPE  \
0  First International Financial Centre-- Bandra ...   Casual Dining   
1                                              Mahim  Dessert Parlor   
2                                               Juhu   Casual Dining   
3                               Bandra Kurla Complex             Bar   
4                               Bandra Kurla Complex             Bar   

            TIMING  RATING_TYPE  RATING  VOTES DAYS_OPENS  
0  12noon to 130am    Excellent     4.9   3529   Mon-Sun)  
1       2pm to 1am    Very Good     4.4   1723   Mon-Sun)  
2    12noon to 1am    Very Good     4.2    337   Mon-Sun)  
3  12noon to 130am    Very Good     4.3   1200   Mon-Sun)  
4    1130am to 1am  Veľmi dobré     4.4   5995   Mon-Sun)

In [34]:
## Good now we are going to remove the brackets from days column 
raw_df['DAYS_OPENS']=raw_df['DAYS_OPENS'].str.replace(')','',regex=True)

In [35]:
raw_df.head()

NAME  PRICE                                    CUSINE_CATEGORY  \
0         Hitchki   1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda    400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu   1800                                      Asian,Chinese   
3  Butterfly High   1000                                      Modern Indian   
4       BKC  DIVE   1200                   North Indian,Chinese,Continental   

                                              REGION     CUSINE TYPE  \
0  First International Financial Centre-- Bandra ...   Casual Dining   
1                                              Mahim  Dessert Parlor   
2                                               Juhu   Casual Dining   
3                               Bandra Kurla Complex             Bar   
4                               Bandra Kurla Complex             Bar   

            TIMING  RATING_TYPE  RATING  VOTES DAYS_OPENS  
0  12noon to 130am    Excellent     4.9   3529    Mon-Sun  
1       2pm to 1am    Very Good     4.4   1723    Mon-Sun  
2    12noon to 1am    Very Good     4.2    337    Mon-Sun  
3  12noon to 130am    Very Good     4.3   1200    Mon-Sun  
4    1130am to 1am  Veľmi dobré     4.4   5995    Mon-Sun

In [36]:
## Checking for null records in Days_Open column
raw_df['DAYS_OPENS'].isnull().sum()

160

In [37]:
raw_df.isnull().sum()

NAME                 0
PRICE                0
CUSINE_CATEGORY      0
REGION               0
CUSINE TYPE          0
TIMING               0
RATING_TYPE          0
RATING               0
VOTES                0
DAYS_OPENS         160
dtype: int64

In [38]:
## we have to replace this 160 No entries values with NA
raw_df.fillna('NA',inplace=True)

In [39]:
## check for dtypes for all columns
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14138 entries, 0 to 15079
Data columns (total 10 columns):
NAME               14138 non-null object
PRICE              14138 non-null int64
CUSINE_CATEGORY    14138 non-null object
REGION             14138 non-null object
CUSINE TYPE        14138 non-null object
TIMING             14138 non-null object
RATING_TYPE        14138 non-null object
RATING             14138 non-null float64
VOTES              14138 non-null int64
DAYS_OPENS         14138 non-null object
dtypes: float64(1), int64(2), object(7)
memory usage: 1.2+ MB


In [40]:
## Removing the restaruant records whose Rating or votes is "0"
## Finding those restaruant whose has "0" as Rating or votes.
useless_data = (raw_df['RATING'] == 0.0) | (raw_df['VOTES'] == 0)  
# | means “either condition can be true.”
raw_df[useless_data]

NAME  PRICE  \
32     Hotel Annapoorna Refreshments    400   
34                         Biryani 9    600   
36                 D Fusion Flavours    350   
39                      Nation Tadka    400   
83              Link  Way Restaurant    500   
...                              ...    ...   
14998                  Foodies House      0   
14999                       Khansama      0   
15010        Earth Cafe @ Waterfield    800   
15023           How About Some Cream    200   
15046          Food And Taste Theory    800   

                                   CUSINE_CATEGORY  \
32                   Maharashtrian,Mughlai,Chinese   
34                            Biryani,North Indian   
36                                         Chinese   
39     North Indian,South Indian,Chinese,Fast Food   
83                            North Indian,Chinese   
...                                            ...   
14998                                      Chinese   
14999                                      Biryani   
15010    Cafe,Healthy Food,Italian,Pizza,Beverages   
15023                                    Beverages   
15046                          Continental,Italian   

                           REGION    CUSINE TYPE                      TIMING  \
32                       Ghansoli    Quick Bites            1030am to 1230AM   
34      Near Andheri East Station           none                 11am to 3am   
36                  Goregaon East           none  12noon to 330pm,7pm to 3am   
39                          Worli           none            12noon to 1230AM   
83                     Jogeshwari    Quick Bites    12noon to 4pm,8pm to 1am   
...                           ...            ...                         ...   
14998               Goregaon East           none               12noon to 4am   
14999                 Lower Parel           none               12noon to 3am   
15010  Linking Road-- Bandra West           Café                10am to 10pm   
15023              Mumbai Central  Beverage Shop               12noon to 3am   
15046  Phoenix Marketcity-- Kurla  Casual Dining           9am to 12midnight   

      RATING_TYPE  RATING  VOTES                     DAYS_OPENS  
32      Not rated     0.0      0                        Mon-Sun  
34           None     0.0      0                        Mon-Sun  
36           None     0.0      0                        Mon-Sun  
39      Not rated     0.0      0                        Mon-Sun  
83      Not rated     0.0      0                        Mon-Sun  
...           ...     ...    ...                            ...  
14998        None     0.0      0                        Mon-Sun  
14999        None     0.0      0                        Mon-Sun  
15010        None     0.0      0   Mon-Thu,10am to 11pm(Fri-Sun  
15023        None     0.0      0                        Mon-Sun  
15046        None     0.0      0  Mon-Fri,9am to 1230AM(Sat-Sun  

[3371 rows x 10 columns]

In [41]:
raw_df["RATING"].value_counts()

0.0    3370
3.5    1094
3.4    1036
3.6     960
3.3     926
3.7     917
3.2     801
3.8     782
3.1     734
3.0     622
3.9     596
2.9     409
4.0     408
2.8     309
4.1     298
4.2     199
2.7     170
4.3     148
4.4      99
2.6      77
4.5      46
2.5      39
4.6      32
2.4      26
4.7      13
2.3      10
2.1       5
2.2       4
4.8       4
4.9       2
2.0       1
1.8       1
Name: RATING, dtype: int64

In [42]:
## Performing negation of the useless dataset and then sorting the correct data back in the raw_df
 ## this permanently remove the useless data from the origional dataframe

raw_df=raw_df[~useless_data]

In [43]:
raw_df

NAME  PRICE  \
0                Hitchki   1200   
1           Baba Falooda    400   
2          Chin Chin Chu   1800   
3         Butterfly High   1000   
4              BKC  DIVE   1200   
...                  ...    ...   
15075    Tirupati Balaji    500   
15076  Hari Om Snack Bar    350   
15077           PitaBurg    400   
15078     Uncha Otlawala    300   
15079     Mandarin Panda    400   

                                         CUSINE_CATEGORY  \
0      Modern Indian,North Indian,Chinese,Momos,Birya...   
1                           Desserts,Ice Cream,Beverages   
2                                          Asian,Chinese   
3                                          Modern Indian   
4                       North Indian,Chinese,Continental   
...                                                  ...   
15075                     Chinese,Fast Food,North Indian   
15076                     Fast Food,South Indian,Chinese   
15077                                 Fast Food,Lebanese   
15078                                 Desserts,Ice Cream   
15079                              Desserts,Chinese,Thai   

                                                  REGION     CUSINE TYPE  \
0      First International Financial Centre-- Bandra ...   Casual Dining   
1                                                  Mahim  Dessert Parlor   
2                                                   Juhu   Casual Dining   
3                                   Bandra Kurla Complex             Bar   
4                                   Bandra Kurla Complex             Bar   
...                                                  ...             ...   
15075                            Oshiwara-- Andheri West   Casual Dining   
15076                                     Kandivali West     Quick Bites   
15077                                        Lower Parel            none   
15078                                     Kandivali West  Dessert Parlor   
15079                                         Malad West            none   

                                TIMING  RATING_TYPE  RATING  VOTES  \
0                      12noon to 130am    Excellent     4.9   3529   
1                           2pm to 1am    Very Good     4.4   1723   
2                        12noon to 1am    Very Good     4.2    337   
3                      12noon to 130am    Very Good     4.3   1200   
4                        1130am to 1am  Veľmi dobré     4.4   5995   
...                                ...          ...     ...    ...   
15075  8am to 11pm,12midnight to 115am         Good     3.5    267   
15076                    11am to 230am         Good     3.7     64   
15077                     11am to 11pm      Average     3.4     99   
15078                    9am to 1230AM         Good     3.5     29   
15079       12noon to 330pm,7pm to 1am         Good     3.7    121   

                            DAYS_OPENS  
0                              Mon-Sun  
1                              Mon-Sun  
2                              Mon-Sun  
3                              Mon-Sun  
4                              Mon-Sun  
...                                ...  
15075                          Mon-Sun  
15076                          Mon-Sun  
15077  Mon,Tue,Wed,Thu,Sun,11am to ...  
15078                          Mon-Sun  
15079                          Mon-Sun  

[10767 rows x 10 columns]

In [44]:
## Working on 'Rating type' column
## checking the unique values in the column 
raw_df['RATING_TYPE'].value_counts()

Average          5111
Good             4330
Very Good        1137
Excellent          95
Poor               47
Veľmi dobré         6
Dobrze              4
Skvělá volba        4
Bardzo dobrze       3
Promedio            2
Ortalama            2
Muito Bom           2
İyi                 2
Bom                 2
Bueno               2
Průměr              2
Dobré               2
Priemer             2
Buono               2
Velmi dobré         1
Biasa               1
Excelente           1
Muy Bueno           1
Çok iyi             1
Sangat Baik         1
Média               1
Skvělé              1
Baik                1
Media               1
Name: RATING_TYPE, dtype: int64

In [45]:
## In the above cell we got Unknown Name so we have to translate it to create visuvalization or analysis.
## Translating the texts into proper English text.
raw_df['RATING_TYPE'].replace(to_replace='Excelente' , value='Excellent', inplace=True) 
raw_df['RATING_TYPE'].replace(to_replace=['Veľmi dobré','Bardzo dobrze','Muy Bueno','Velmi dobré'] , value='Very Good',inplace=True)
raw_df['RATING_TYPE'].replace(to_replace=['Skvělá volba','Dobrze','Bueno','Buono','Dobré','Bom','Skvělé'] , value='Good', inplace=True)
raw_df['RATING_TYPE'].replace(to_replace=['Priemer','Média','Çok iyi'] , value='Average', inplace=True) 
raw_df['RATING_TYPE'].replace(to_replace=['Průměr','Promedio','Ortalama','Muito Bom','İyi'] , value='Poor', inplace=True)
raw_df['RATING_TYPE'].replace(to_replace=['Baik','Biasa','Media','Sangat Baik'] , value='Very Poor', inplace=True) 

In [46]:
## Check for correction (Translation)
raw_df['RATING_TYPE'].value_counts()

Average      5115
Good         4347
Very Good    1148
Excellent      96
Poor           57
Very Poor       4
Name: RATING_TYPE, dtype: int64

In [47]:
## Working on 'Regeion' column 
raw_df['REGION'].value_counts()

Mira Road                             405
Malad West                            308
Chembur                               277
Kharghar                              268
Borivali West                         264
                                     ... 
Fariyas Hotel-- Colaba                  1
Royal Challenge Complex-- Goregaon      1
Sofitel-- Bandra Kurla Complex          1
Yogi Midtown-- Turbhe                   1
Comfort Inn Heritage-- Byculla          1
Name: REGION, Length: 237, dtype: int64

## 'Hotel Jewel of Chembur-- Chembur' In this record we want only region name not hotel name 
## we should remove unwanted text in region column 
# How we remove unwanted text using regax (regular expression)
## Regax decides what to keep,remove, or replace.

In [48]:
## Removing the irrelevant text from the region column 
raw_df['REGION'] = raw_df['REGION'].str.replace('[a-zA-Z].+-- ','',regex=True) 
#raw_df["REGION"]=raw_df["REGION"].str.replace('[a-zA-Z].+--', regex=True)

In [49]:
## Removing the west and east from the region column
raw_df['REGION']=raw_df["REGION"].str.replace('West|west|East|east','',regex=True)

In [50]:
raw_df['REGION'].value_counts()  # Irrelevent and unwanted texts are removed.

Thane                726
Mira Road            412
Andheri              409
Malad                378
Kandivali            377
                    ... 
Kalyan                 2
CBD Belapur            1
Goregaon               1
Girgaon Chowpatty      1
Dadar                  1
Name: REGION, Length: 104, dtype: int64

In [51]:
## Replacing small regions with know region names (know region already listed in region column)
raw_df['REGION'] = raw_df['REGION'].str.replace('4 Bungalows|7 Andheri|Azad Nagar|Near Andheri Station|Veera Desai Area','Andheri',regex=True)
raw_df['REGION'] = raw_df['REGION'].str.replace('Bandra Kurla Complex','Bandra',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('CBD-Belapur','CBD Belapur',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Girgaon Chowpatty','Chowpatty',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Dadar Shivaji Park','Dadar',regex=True)
raw_df['REGION'] = raw_df['REGION'].str.replace('Flea Bazaar Café|Kamala Mills Compound','Lower Parel',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Runwal Green','Mulund',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Mumbai CST Area','Mumbai Central',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Kopar Khairane|Seawoods|Turbhe|Ulwe','Navi Mumbai',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('New Panvel|Old Panvel','Panvel',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Kamothe','Sion',regex=True) 
raw_df['REGION'] = raw_df['REGION'].str.replace('Ghodbunder Road|Majiwada','Thane',regex=True)

In [52]:
raw_df.head()

NAME  PRICE                                    CUSINE_CATEGORY  \
0         Hitchki   1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda    400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu   1800                                      Asian,Chinese   
3  Butterfly High   1000                                      Modern Indian   
4       BKC  DIVE   1200                   North Indian,Chinese,Continental   

   REGION     CUSINE TYPE           TIMING RATING_TYPE  RATING  VOTES  \
0  Bandra   Casual Dining  12noon to 130am   Excellent     4.9   3529   
1   Mahim  Dessert Parlor       2pm to 1am   Very Good     4.4   1723   
2    Juhu   Casual Dining    12noon to 1am   Very Good     4.2    337   
3  Bandra             Bar  12noon to 130am   Very Good     4.3   1200   
4  Bandra             Bar    1130am to 1am   Very Good     4.4   5995   

  DAYS_OPENS  
0    Mon-Sun  
1    Mon-Sun  
2    Mon-Sun  
3    Mon-Sun  
4    Mon-Sun

In [53]:
## Removing duplicate records (Repeated names)
## Finding all the duplicate rows.
raw_df[raw_df.duplicated()] #It shows only duplicate same name,price everything so, we have to remove this duplicate records.

NAME  PRICE  \
4064   Sai Sannidhi Restaurant & Bar   1000   
4068                    Konkan Katta    400   
4082                  Usmaniya Hotel    600   
4083                    Gina's Cakes    450   
4084          Konkanastha Lunch Home    400   
...                              ...    ...   
14200      Mezbaan Family Restaurant    350   
14204               Jyoti Lunch Home    650   
14253                        On Toes    900   
14761                    Frosty Farm    400   
14928               Shree Manu Sagar    300   

                            CUSINE_CATEGORY      REGION     CUSINE TYPE  \
4064                    North Indian,Konkan    Dahisar    Casual Dining   
4068          Seafood,Maharashtrian,Malwani    Mahakali     Quick Bites   
4082                                Mughlai        Fort   Casual Dining   
4083                                 Bakery  Dombivali             none   
4084                        Seafood,Malwani     Chakala   Casual Dining   
...                                     ...         ...             ...   
14200                       Chinese,Mughlai      Mumbra           Dhaba   
14204  Chinese,North Indian,Seafood,Mughlai     Mulund    Casual Dining   
14253          Italian,North Indian,Chinese      Malad    Casual Dining   
14761          Ice Cream,Desserts,Fast Food      Malad   Dessert Parlor   
14928           North Indian,Chinese,Indian    Ghansoli     Quick Bites   

                              TIMING RATING_TYPE  RATING  VOTES DAYS_OPENS  
4064              11am to 12midnight        Good     3.7     99    Mon-Sun  
4068   11am to 330pm,630pm to 1130pm        Good     3.5    181    Mon-Sun  
4082                1030am to 1130pm     Average     3.2      8    Mon-Sun  
4083                    11am to 11pm        Good     3.5     49    Mon-Sun  
4084   12noon to 3pm,730pm to 1030pm        Good     3.5     44    Mon-Sun  
...                              ...         ...     ...    ...        ...  
14200               12noon to 1230AM     Average     2.8     97    Mon-Sun  
14204                 11am to 1230AM        Good     3.5     49    Mon-Sun  
14253    12noon to 3pm,7pm to 1230AM        Good     3.6     76    Mon-Sun  
14761                  1pm to 1215AM        Good     3.6    120    Mon-Sun  
14928  1130am to 415pm,7pm to 1215AM     Average     3.3     51    Mon-Sun  

[220 rows x 10 columns]

In [54]:
## droping all the duplicate rows.
raw_df=raw_df.drop_duplicates()

In [55]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10547 entries, 0 to 15079
Data columns (total 10 columns):
NAME               10547 non-null object
PRICE              10547 non-null int64
CUSINE_CATEGORY    10547 non-null object
REGION             10547 non-null object
CUSINE TYPE        10547 non-null object
TIMING             10547 non-null object
RATING_TYPE        10547 non-null object
RATING             10547 non-null float64
VOTES              10547 non-null int64
DAYS_OPENS         10547 non-null object
dtypes: float64(1), int64(2), object(7)
memory usage: 906.4+ KB


In [56]:
## coping the cleaned dataset into a new dataframe
zomato_df=raw_df.copy()
zomato_df.head()

NAME  PRICE                                    CUSINE_CATEGORY  \
0         Hitchki   1200  Modern Indian,North Indian,Chinese,Momos,Birya...   
1    Baba Falooda    400                       Desserts,Ice Cream,Beverages   
2   Chin Chin Chu   1800                                      Asian,Chinese   
3  Butterfly High   1000                                      Modern Indian   
4       BKC  DIVE   1200                   North Indian,Chinese,Continental   

   REGION     CUSINE TYPE           TIMING RATING_TYPE  RATING  VOTES  \
0  Bandra   Casual Dining  12noon to 130am   Excellent     4.9   3529   
1   Mahim  Dessert Parlor       2pm to 1am   Very Good     4.4   1723   
2    Juhu   Casual Dining    12noon to 1am   Very Good     4.2    337   
3  Bandra             Bar  12noon to 130am   Very Good     4.3   1200   
4  Bandra             Bar    1130am to 1am   Very Good     4.4   5995   

  DAYS_OPENS  
0    Mon-Sun  
1    Mon-Sun  
2    Mon-Sun  
3    Mon-Sun  
4    Mon-Sun

## Performing exploratory data analysis
## Q1) How many restaurants are in Mumbai for each type of cuisine?

## Exploratory means _ both descriptive and inferential analysis 

In [57]:
! pip install -U kaleido